# Background

This folder contains single-ended V4 16S rRNA sequencing from 3 pregnant women, sampled at 3 body sites (fecal, oral, and vaginal), at 3 time points (second trimester, third trimester, and shortly after birth). You will find the reads within `reads.tar.gz`. You will also find relevant metadata within metadata.xlsx which includes a tab with variable explanations (vars). Your assignment will be based around the analysis of this data set to *answer questions about variation in the microbiome across body sites*. Your final written report may be provided either as a PDF in the traditional course format, or as a knit markdown document. If provided as a markdown document, it should be still be referenced and formatted in a fashion suitable for a formal document

# Questions

For each question below, be sure to include *at least one* visualization and/or statistical analysis to support your answer and include references to the literature that support or contrast with your findings.

-[ ] Process this raw sequencing data using dada2 (Note: this is single ended data which will require some slight modifications to the workflow, but requires less computational power). See documentation [here](https://benjjneb.github.io/dada2/). Describe your work flow.
-[ ] Is this sequencing data of high quality? Provide data to support your answer.
-[ ] How does alpha diversity vary across body sites?
-[ ] How does beta diversity vary across body sites?
-[ ] Is body site or inter-individual variation a greater driver of microbiome composition?
-[ ] What are the major differences in genus-level abundances between body sites?
-[ ] Where in the world do you think these women were located and why? Do you think microbiome composition would be a reliable marker of geography?
-[ ] Pick one additional variable from the metadata for analysis of the *gut microbiota*. How does microbial diversity and taxonomic composition vary with this variable?